torchtext/experimental/transforms.py

```diff
@@ -340,3 +341,144 @@ def forward(self, tokens: List[str]) -> Tensor:
+                mask_prob = 0.15
+            )
+        >>> masked_tokens, target_tokens, mask = mask_transform(sample_token_ids)
+    """
+
+    # maks_mask_prob is prob. of replacing a token with [MASK] (ex. 80%)
+    mask_mask_prob = 0.8
+
+    # rand_mask_thresh is prob. of replacing a token with a random token. (ex.10%)
+    rand_mask_prob = 0.1
+
+    def __init__(
+        self,
+        vocab_len: int,
+        mask_idx: int,
+        bos_idx: int,
+        pad_idx: int,
+        mask_bos: bool = False,
+        mask_prob: float = 0.15,
+    ):
+        super().__init__()
+        self.vocab_len = vocab_len
+        self.mask_idx = mask_idx
+        self.bos_idx = bos_idx
+        self.pad_idx = pad_idx
+        self.mask_prob = mask_prob
+        self.mask_bos = mask_bos
+
+    def forward(self, tokens: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """
+        Applies mask to input tokens.
+
+        Inputs:
+            tokens: Tensor with token ids of shape (batch_size x seq_len). Includes token ids for special tokens such as [BOS] and [PAD]
+
+        Outputs:
+            masked_tokens: Tensor of tokens after masking has been applied
+            target_tokens: Tensor of token values selected for masking
+            mask: Tensor with same shape as input tokens (batch_size x seq_len)
+                with masked tokens represented by a 1 and everything else as 0.
+        """
+        # tokens, mask, mask_mask, rand_mask: (T, C)
+        mask, mask_mask, rand_mask = self._generate_mask(tokens)
+
+        # a. generate the masked input tokens
+        # (1) the [MASK] token 80% of the time
+        masked_tokens = self._mask_input(tokens, mask_mask, self.mask_idx)
+        # (2) a random token 10% of the time
+        masked_tokens = self._mask_input(
+            masked_tokens,
+            rand_mask,
+            torch.randint_like(tokens, high=self.vocab_len),
+        )
+
+        # b. generate the target prediction
+        target_tokens = torch.masked_select(tokens, mask.bool())
+
+        # masked_tokens: (T, C), target_tokens: (T x C x mask_prob, ), mask
+        return masked_tokens, target_tokens, mask
+
+    def _random_masking(self, tokens: torch.tensor, mask_prob: float) -> torch.Tensor:
+        """
+        Function to mask tokens randomly.
+
+        Inputs:
+            1) tokens: Tensor with token ids of shape (batch_size x seq_len)
+            2) mask_prob: Probability of masking a particular token
+
+        Outputs:
+            mask: Tensor with same shape as input tokens (batch_size x seq_len)
+                with masked tokens represented by a 1 and everything else as 0.
+        """
+        batch_size, seq_len = tokens.size()
+        num_masked_per_seq = int(seq_len * mask_prob)
+
+        mask = np.zeros((batch_size, seq_len), dtype=np.int_)
+        mask[:, :num_masked_per_seq] = 1
+        for row in mask:
+            np.random.shuffle(row)
```

Comment:
Can we replace NumPy with PyTorch?

This brings in NumPy dependency and it uses NumPy's global RNG, which makes it harder to control the randomness of the code.